lib/url.c

```diff
@@ -1668,18 +1675,28 @@ static struct connectdata *allocate_conn(struct Curl_easy *data)
      data becomes proxy backend data). */
   {
     size_t sslsize = Curl_ssl->sizeof_ssl_backend_data;
-    char *ssl = calloc(SSL_BACKEND_CNT, sslsize);
-    if(!ssl) {
+    bool allocate_result = true;
+
+    conn->ssl[FIRSTSOCKET].backend = calloc(1, sslsize);
+    allocate_result &= conn->ssl[FIRSTSOCKET].backend != NULL;
+#ifndef CURL_DISABLE_FTP
+    conn->ssl[SECONDARYSOCKET].backend = calloc(1, sslsize);
+    allocate_result &= conn->ssl[SECONDARYSOCKET].backend != NULL;
+#endif
+#ifndef CURL_DISABLE_PROXY
+    conn->proxy_ssl[FIRSTSOCKET].backend = calloc(1, sslsize);
+    allocate_result &= conn->proxy_ssl[FIRSTSOCKET].backend != NULL;
+#ifndef CURL_DISABLE_FTP
+    conn->proxy_ssl[SECONDARYSOCKET].backend = calloc(1, sslsize);
+    allocate_result &= conn->proxy_ssl[SECONDARYSOCKET].backend != NULL;
+#endif
+#endif
```

Comment:
I'm not a total fan of how this turns one malloc into four for a regular build. How about instead making the function similar to this? (untested)
~~~c
static struct connectdata *allocate_conn(struct Curl_easy *data)
{
  struct connectdata *conn = calloc(1, sizeof(struct connectdata));
  if(!conn)
    return NULL;

#ifdef USE_SSL
  /* The SSL backend-specific data (ssl_backend_data) objects are allocated as
     a separate array to ensure suitable alignment.
     Note that these backend pointers can be swapped by vtls (eg ssl backend
     data becomes proxy backend data). */
  {
    size_t onesize = Curl_ssl->sizeof_ssl_backend_data;
    size_t totalsize = onesize;
    void *ptr;
    bool allocate_result = true;

#ifndef CURL_DISABLE_FTP
    totalsize *= 2;
#endif
#ifndef CURL_DISABLE_PROXY
    topalsize *= 2;
#endif

    conn->ssl[FIRSTSOCKET].backend = ptr = calloc(1, totalsize);
    if(!ptr)
      return NULL;
#ifndef CURL_DISABLE_FTP
    conn->ssl[SECONDARYSOCKET].backend = ptr += onesize;
#endif
#ifndef CURL_DISABLE_PROXY
    conn->proxy_ssl[FIRSTSOCKET].backend = ptr += onesize;
#ifndef CURL_DISABLE_FTP
    conn->proxy_ssl[SECONDARYSOCKET].backend = ptr += onesize;
#endif
#endif
  }
#endif

Reply:
I have modified the memory allocation as you wrote.